modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java

```diff
@@ -625,6 +761,117 @@ public synchronized int read() throws IOException {
                                               // we read the input stream (i.e. when it's rate limited)
     }
 
+    private static InputStream toSynchronizedInputStream(String blobName, InputStream delegate, MultiPart multipart) {
+        assert delegate.markSupported() : "An InputStream with mark support was expected";
+        // We need to introduce a read barrier in order to provide visibility for the underlying
```

Comment:
Let us leave this for now, but it seems strange to need this synchronized part really, since we expect it to be used serially only anyway - and this must require a happens-before relationship established in reactor code even if this is used across threads.